internal/configuration/schema/types_test.go

```diff
@@ -48,3 +55,499 @@ func TestNewAddressFromString(t *testing.T) {
+		{"ShouldNotParseEmptyData", "", "", ""},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual, err := NewX509CertificateChain(tc.have)
+
+			switch len(tc.err) {
+			case 0:
+				switch len(tc.have) {
+				case 0:
+					assert.Nil(t, actual)
+				default:
+					assert.NotNil(t, actual)
+
+					assert.Equal(t, tc.thumbprintSHA256, fmt.Sprintf("%x", actual.Thumbprint(crypto.SHA256)))
+					assert.True(t, actual.HasCertificates())
+				}
+				assert.NoError(t, err)
+			default:
+				assert.Nil(t, actual)
+				assert.EqualError(t, err, tc.err)
+			}
+		})
+	}
+}
+
+func TestX509CertificateChain(t *testing.T) {
+	chain := &X509CertificateChain{}
+
+	assert.Nil(t, chain.Thumbprint(crypto.SHA256))
+	assert.False(t, chain.HasCertificates())
+	assert.Len(t, chain.Certificates(), 0)
+
+	assert.False(t, chain.Equal(nil))
+	assert.False(t, chain.Equal(&x509.Certificate{}))
+
+	assert.False(t, chain.EqualKey(nil))
+	assert.False(t, chain.EqualKey(&rsa.PrivateKey{}))
+
+	cert := MustParseCertificate(x509CertificateRSA)
+	cacert := MustParseCertificate(x509CACertificateRSA)
+
+	chain = MustParseX509CertificateChain(x509CertificateRSA + "\n" + x509CACertificateRSA)
+	key := MustParseRSAPrivateKey(x509PrivateKeyRSA)
+
+	thumbprint := chain.Thumbprint(crypto.SHA256)
+	assert.NotNil(t, thumbprint)
+	assert.Equal(t, "956e53c127b18143241fb75d2149369d41ad1e4094c40cedcef22e468b37886b", fmt.Sprintf("%x", thumbprint))
+
+	assert.True(t, chain.Equal(cert))
+	assert.False(t, chain.Equal(cacert))
+	assert.True(t, chain.EqualKey(key))
+
+	assert.NoError(t, chain.Validate())
+
+	chain = MustParseX509CertificateChain(x509CertificateRSA + "\n" + x509CertificateRSA)
+	assert.EqualError(t, chain.Validate(), "certificate #1 in chain is not signed properly by certificate #2 in chain: x509: invalid signature: parent certificate cannot sign this kind of certificate")
+
+	chain = MustParseX509CertificateChain(x509CertificateRSAExpired + "\n" + x509CACertificateRSAExpired)
+
+	err := chain.Validate()
+	require.NotNil(t, err)
+	assert.Regexp(t, regexp.MustCompile(`^certificate #1 in chain is invalid after 31536000 but the time is \d+$`), err.Error())
+
+	chain = MustParseX509CertificateChain(x509CertificateRSANotBefore + "\n" + x509CACertificateRSAotBefore)
+
+	err = chain.Validate()
+	require.NotNil(t, err)
+	assert.Regexp(t, regexp.MustCompile(`^certificate #1 in chain is invalid before 13569465600 but the time is \d+$`), err.Error())
+}
+
+func MustParseX509CertificateChain(data string) *X509CertificateChain {
+	chain, err := NewX509CertificateChain(data)
+	if err != nil {
+		panic(err)
+	}
+
+	if chain == nil {
+		panic("nil chain")
+	}
+
+	return chain
+}
+
+func MustParseCertificate(data string) *x509.Certificate {
+	block, x := pem.Decode([]byte(data))
+	if block == nil {
+		panic("not pem")
+	}
+
+	if len(x) != 0 {
+		panic("extra data")
+	}
+
+	if block.Type != "CERTIFICATE" {
```

Comment:
```suggestion
	if block.Type != certificate {
```

Can change this if we move the const.